### PR TITLE
DVDInterface: Read disc after delay, not before

### PR DESCRIPTION
**Pull request by JosJuice**
DVDInterface behavior before 4.0-4222:
* Receive and parse command
* Wait for a while to simulate loading times
* Read data from disc image and copy to emulator memory
* Reply to command

DVDInterface behavior starting with 4.0-4222:
* Receive and parse command
* Read data from disc image and copy to emulator memory
* Wait for a while to simulate loading times
* Reply to command

The former is obviously more accurate, because a real console isn't able to copy all the data to RAM before it has finished reading it from the disc. I changed it in order to not make the code unnecessarily complicated when adding a few things... and games shouldn't be reading the memory that is being copied to, right? Unfortunately, Resident Evil 3 does exactly that, and plays incorrect audio starting with 4.0-2222: https://code.google.com/p/dolphin-emu/issues/detail?id=8026

The main obstacle to reverting the order things happen in is that WII_IPC_HLE only can schedule a "reply to command" event, not arbitrary things like "read disc data and reply to command" that we would need. Because of this, WII_IPC_HLE_Device_DI::IOCtl now returns false and lets DVDInterface::ExecuteCommand schedule an event instead. A second obstacle is that the scheduled event needs to know what data to read. Since DVDInterface now supports WII_IPC_HLE_Device_DI, we can't just read the hardware registers like before 4.0-4222. We can't use pointers in the event userdata either, because it would break savestates. The solution I used is a deque, where a function looks up commands based on an ID that is passed as userdata.

Most of this solution is actually based on my plans for making a separate thread for disc IO. That has additional reasons for not letting WII_IPC_HLE schedule reads, mainly that it would be nice to be able to finish reads right after the disc IO thread finishes reading from the disc image, but I think this solution is OK for the current requirements too. The good thing is that most of the code in this PR can be used without any change when/if a disc IO thread is made, instead of having to write something similar again.

Because this touches WII_IPC_HLE timing, chances are that it will break stuff.

### CODE REVIEW

File: Source/Core/Core/State.cpp [hunk 1]
```diff
@@ -64,7 +64,7 @@ static Common::Event g_compressAndDumpStateSyncEvent;
 static std::thread g_save_thread;
 
 // Don't forget to increase this after doing changes on the savestate system
-static const u32 STATE_VERSION = 38;
+static const u32 STATE_VERSION = 40;
```
Comment (JosJuice) [inline Source/Core/Core/State.cpp]: Should I increment STATE_VERSION twice like this because two commits introduce savestate differences, or should it only be once for the entire PR?

**Comment by RachelBryk** [comment]
>Because this touches WII_IPC_HLE, chances are that it will break stuff. And indeed, things are broken right now. If you start and exit any game, then try to start a Wii game, the game will freeze, and it seems like Dolphin freezes too when you try to exit the game. I don't know the cause yet, but it needs to be fixed before the WIP label is removed from this PR.

Looks like static variables are used and not reset when emulation starts/ends? Just skimmed through, but that's probably it.

File: Source/Core/Core/State.cpp (hunk 1 above)
Comment (RachelBryk) [inline Source/Core/Core/State.cpp]: once

**Comment by JosJuice** [comment]
Oh right, ioctl_callback in WII_IPC_HLE_Device_DI needs to be re-initialized every time emulation starts. I'll fix it.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp [hunk 2]
```diff
@@ -63,14 +78,17 @@ IPCCommandResult CWII_IPC_HLE_Device_di::IOCtl(u32 _CommandAddress)
...
 
-	if (SConfig::GetInstance().m_LocalCoreStartupParameter.bFastDiscSpeed)
-		result.ticks_until_completion = 0;	// An optional hack to speed up loading times
+	// DVDInterface's ExecuteCommand handles timing, disc reading... basically everything
+	ExecuteCommand(command_0, command_1, command_2, BufferOut, BufferOutSize,
+	               false, ioctl_callback, _CommandAddress);
 
-	return { true, result.ticks_until_completion };
+	// The reply is sent by IOCtlCallback instead of WII_IPC_HLE_Interface
+	return IPC_NO_REPLY;
```
Comment (magumagu) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: Isn't this going to mess with the timing of the reply?  You're ignoring the normal reply queue.

Comment (JosJuice) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: Yes. That's the most problematic thing with this PR in my opinion. I think it makes some sense that IOS would be able to execute other IPC commands while waiting for the disc drive execute a DI command, but I'm not sure if it's correct.

File: Source/Core/Core/HW/DVDInterface.cpp [hunk 3]
```diff
@@ -257,8 +281,8 @@ void UpdateInterrupts();
 void GenerateDIInterrupt(DIInterruptType _DVDInterrupt);
 
 void WriteImmediate(u32 value, u32 output_address, bool write_to_DIIMMBUF);
-DVDCommandResult ExecuteReadCommand(u64 DVD_offset, u32 output_address,
-	u32 DVD_length, u32 output_length, bool decrypt);
+DVDReadCommand ExecuteReadCommand(u64 DVD_offset, u32 output_address, u32 DVD_length, u32 output_length,
+                                  bool decrypt, DIInterruptType* interrupt_type, u64* ticks_until_completion);
```
Comment (Buddybenj) [inline Source/Core/Core/HW/DVDInterface.cpp]: I think Dolphin uses tabs at the beginning and spaces at the end to line it up. I may be wrong though.

Comment (FioraAeterna) [inline Source/Core/Core/HW/DVDInterface.cpp]: I think that's correct.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.h [hunk 4]
```diff
@@ -25,4 +25,8 @@ class CWII_IPC_HLE_Device_di : public IWII_IPC_HLE_Device
 
 	IPCCommandResult IOCtl(u32 _CommandAddress) override;
 	IPCCommandResult IOCtlV(u32 _CommandAddress) override;
+
+private:
+
```
Comment (Buddybenj) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.h]: Extraneous new line.

**Comment by tueidj** [comment]
WII_IPC based "cooked" (decrypted) reads still take longer than they should, due to https://code.google.com/p/dolphin-emu/issues/detail?id=7890
Related: decryption shouldn't be implemented at the disk interface level - the drive hardware doesn't know anything about it, it's handled by IOS and moving it to WII_IPC_HLE_Device_DI would make sure the disc transfer rate is only applied to fetching the data, not decrypting it (or erroneously applying it to an already decrypted block).

**Comment by JosJuice** [comment]
Moving the decryption to WII_IPC_HLE_Device_DI would be cleaner, but it wouldn't work because of VolumeDirectory. It kind of ugligly inserts itself as a Volume instead of a FileBlob so that it can read unencrypted files and pretend that they are a disc image that has been decrypted. It would be nice to do something about VolumeDirectory, but I don't know what. Anyway, it isn't really related to this PR.

**Comment by comex** [comment]
It might be nice to start the actual host disk read (asynchronously) before the delay, to prevent stuttering.  Although I have no idea whether such stuttering actually occurs.

**Comment by JosJuice** [comment]
That's the purpose of the disc IO thread I mentioned.

**Comment by Buddybenj** [comment]
Bump.

**Comment by JosJuice** [comment]
It turns out that the WII_IPC_HLE timing change in this PR fixes a regression from the previous time I changed the timing (4.0-4898). https://code.google.com/p/dolphin-emu/issues/detail?id=8158

File: Source/Core/Core/HW/DVDInterface.cpp [hunk 5]
```diff
@@ -301,6 +329,49 @@ static void TransferComplete(u64 userdata, int cyclesLate)
...
+	else
+	{
+		// Here is the actual disc reading
+		if (!DVDRead(command.DVD_offset, command.output_address, command.length, command.decrypt))
+			PanicAlertT("Can't read from DVD_Plugin - DVD-Interface: Fatal Error");
+	}
+
+	// The final step is to notify the emulated software that the command has been executed
+	// There isn't a ScheduleEvent_Immediate, so we use ScheduleEvent_Threadsafe_Immediate
+	CoreTiming::ScheduleEvent_Threadsafe_Immediate(command.callback_event_type, command.callback_userdata);
```
Comment (magumagu) [inline Source/Core/Core/HW/DVDInterface.cpp]: Why is this using a CoreTiming event as opposed to a function pointer?

Comment (JosJuice) [inline Source/Core/Core/HW/DVDInterface.cpp]: I thought pointers don't work with savestates? Or does that only apply when having a pointer in a CoreTiming event's userdata? Either way, the pointer/event we use here is the one that has been passed to ExecuteCommand earlier, and the final part of ExecuteCommand requires a CoreTiming event in case the command isn't a read command, so the ExecuteCommand argument can't be changed to a pointer.

Comment (magumagu) [inline Source/Core/Core/HW/DVDInterface.cpp]: The issue with savestates would be that you can't store a function pointer in DVDReadCommand?  I guess that's correct. It still seems like you're abusing the CoreTiming interface, though: ScheduleEvent_Immediate doesn't exist because CoreTiming isn't supposed to be used as just an int->function pointer table.

Maybe instead of making ExecuteCommand take a callback_event_type as a parameter, make it not interact directly with CoreTiming at all.  It can just return the number of cycles to wait before calling FinishExecuteReadCommand, and then the callers can explicitly handle scheduling an event and calling FinishExecuteReadCommand when the read finishes.  That seems much more straightforward in terms of describing what code runs when.

File: Source/Core/Core/HW/DVDInterface.cpp [hunk 6]
```diff
@@ -236,11 +255,16 @@ static u32  g_ErrorCode = 0;
...
+static int  finish_execute_command = 0;
+static int  finish_execute_read_command = 0;
 static int  dtk = 0;
 
 static u64 g_last_read_offset;
 static u64 g_last_read_time;
+static u64 g_last_command_time;
+
+static std::deque<DVDReadCommand> read_commands_to_execute;
+static u64 next_read_command_id;
```
Comment (magumagu) [inline Source/Core/Core/HW/DVDInterface.cpp]: Why do we have a queue here?  As far as I know, it isn't possible to send multiple read commands to the DVD drive at the same time.  (It's possible WII_IPC_HLE_Device_DI needs a queue, but that shouldn't be here.)

**Comment by magumagu** [comment]
re: issue 8158: it's nice to have some confirmation IOS can in fact handle multiple commands at the same time.

**Comment by magumagu** [comment]
Also, sorry about the delayed review; I didn't really notice those issues the first time I looked at this.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp [hunk 7]
```diff
@@ -563,6 +563,11 @@ void EnqueueReply_Threadsafe(u32 address, int cycles_in_future)
 	CoreTiming::ScheduleEvent_Threadsafe(cycles_in_future, event_enqueue, address);
 }
 
+void EnqueueReply_Immediate(u32 address)
+{
+	EnqueueEventCallback(address, 0);
+}
```
Comment (magumagu) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp]: It would be nice if you could refactor this so it calls something with a more relevant name; it isn't really intuitive why you're calling a function named EnqueueEventCallback, and why you're passing it zero.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp (hunk 2 above)
Comment (magumagu) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: If you get rid of the last_reply_time hack in WII_IPC_HLE.cpp, independently scheduling the reply is fine, because it doesn't change behavior.  Might be worth looking into.

File: Source/Core/Core/HW/DVDInterface.cpp (hunk 5 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDInterface.cpp]: Yeah, this isn't really what CoreTiming should be used for... But I don't think there are any better options. The reason that I don't just return the number of cycles to wait is that it would lead to code duplication. Both DVDInterface and Device_DI would need to handle scheduling both read commands and other commands. Also, in the future, there are things that wouldn't be possible to add if ExecuteCommand needs to expose the delay to callers. For instance, with the disc IO thread, it would be nice to have an option to call FinishExecuteReadCommand (and the callback) right after the host read finishes, and that would make it impossible to know the delay in advance.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp (hunk 2 above)
Comment (JosJuice) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: I don't think I understand... If we didn't have last_reply_time, what would we have instead? It sounds unlikely that things would work if we have complete anarchy with the timings.

File: Source/Core/Core/HW/DVDInterface.cpp (hunk 6 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDInterface.cpp]: It's theoretically possible to send a DI command before the last one is done, but that doesn't make any sense, so it's not really something we have to care about. So, hmm... There could be a queue in Device_DI instead, where incoming IPC commands wait until DVDInterface isn't executing anything. That would also make it possible to get rid of g_last_command_time from DVDInterface, so it sounds like a good idea in overall.

File: Source/Core/Core/HW/DVDInterface.cpp (hunk 5 above)
Comment (magumagu) [inline Source/Core/Core/HW/DVDInterface.cpp]: Well, if you really need to structure the code this way, you could pass in an enum, and have FinishExecuteReadCommand explicitly call either some DVDInterface function or a WII_IPC_HLE function.  But I guess that's pretty similar to CoreTiming.

If you are going to do this, just add ScheduleEvent_Immediate.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp (hunk 2 above)
Comment (magumagu) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: It isn't complete anarchy: even without last_reply_time, the queues for replies and acknowledgements force some ordering onto requests/replies.  And DVD read requests will come out in the right order because that essentially has its own queue.  Any delay based on accessing some particular device (like the DVD drive or NAND) probably needs to queue requests, but I would guess most commands take very little CPU time on Starlet.

Maybe we could add some sort of rough CPU-time approximation, like saying that Starlet can only generate 1 reply every 250 us or whatever, but it doesn't make sense to freeze the reply queue for hundreds of milliseconds because Starlet is waiting for the DVD drive to generate an interrupt.  The system is clearly designed to allow out-of-order replies.

Comment (JosJuice) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: Sounds reasonable. This is something that could use an improvement, but it should be done in another PR, not this one that originally only was intended to change the order of things in DVDInterface.

Comment (magumagu) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: Okay.

**Comment by JosJuice** [comment]
Good news: There is now a queue in WII_IPC_HLE_Device_DI and no queue in DVDInterface, making things a bit cleaner.
Bad news: I don't know how to set a class method as a CoreTiming callback, so I made Device_DI's IOCtlCallback a static function instead of a class method, which made me have to do the same for everything it calls and every variable it accesses... It got quite messy. What can be done about it?

**Comment by magumagu** [comment]
For your callback problem: you can make a global variable like "WII_IPC_HLE_Device_DI* di_pointer" and set up a callback wrapper to do something like "di_pointer->callback();", if that lets you organize your code better.

I'll take a closer look later.

File: Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp [hunk 8]
```diff
@@ -19,13 +20,81 @@
...
+	{
+		// Set out buffer to zeroes as a safety precaution
+		// to avoid answering nonsense values
+		Memory::Memset(BufferOut, 0, BufferOutSize);
+	}
+
+	// The original hardware overwrites the command type with the async reply type.
+	Memory::Write_U32(IPC_REP_ASYNC, command_address);
+	// IOS also seems to write back the command that was responded to in the FD field.
+	Memory::Write_U32(Memory::Read_U32(command_address), command_address + 8);
```
Comment (magumagu) [inline Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp]: Might be a little more clear to move these writes to IOCtlCallback.  (IIRC, it doesn't actually matter when they run, so might as well do all the writing at the same time.)

**Comment by JosJuice** [comment]
Everything should hopefully be fixed now.

**Comment by magumagu** [comment]
LGTM.

**Comment by magumagu** [comment]
I'd appreciate it if this could get a second review.

**Comment by phire** [comment]
LGTM

**Comment by JMC47** [comment]
ping?
